plugins/badges-backend/src/service/router.ts

```diff
@@ -125,25 +187,255 @@ export async function createRouter(
+        .update(`${kind}:${namespace}:${name}:${salt}`)
+        .digest('hex');
+
+      lookupTable.set(hash, { name, namespace, kind, creationDate });
+
+      return lookupTable;
+    });
+
+    // Monitor the lookup generation time and log it
+    logger.info(
+      `Finished generating lookup table in ${
+        Date.now() - generationStartTimestamp
+      } ms`,
+    );
+    // Monitor the lookup table size in byte
+    logger.info(`Lookup table size: ${lookupTable.size} entries`);
+
+    return lookupTable;
+  }
+
+  // This function check if the lookup table is empty or if it is expired (based on the cacheTimeToLive config) and need to be refreshed
+  function isLookupTableToRefresh(
+    table: Map<
+      string,
+      {
+        name: string;
+        namespace: string;
+        kind: string;
+        creationDate: number;
+      }
+    >,
+  ): boolean {
+    const lookupTableLength = table.size;
+
+    if (lookupTableLength === 0) {
+      return true;
+    }
+
+    const now = Date.now();
+    const cacheTimeToLive =
+      options.config.getOptionalNumber(
+        'custom.badges-backend.cacheTimeToLive',
+      ) ?? 10 * 60 * 1000;
+
+    const valuesArray = Array.from(table.values());
+    const lastEntry = valuesArray[valuesArray.length - 1];
+    const lastCreationDate = lastEntry.creationDate;
+
+    return lastCreationDate + cacheTimeToLive < now;
+  }
+
+  // This function return the namespace/kind/name triplet from the lookup table based on the hash
+  async function getEntityInfoFromLookupTable(
+    entityHash: string,
+  ): Promise<{ name: string; namespace: string; kind: string }> {
+    if (lookupTable.get(entityHash) === undefined) {
+      throw new NotFoundError(`No entity with hash "${entityHash}"`);
+    }
+
+    const name = lookupTable.get(entityHash)!.name;
+    const namespace = lookupTable.get(entityHash)!.namespace;
+    const kind = lookupTable.get(entityHash)!.kind;
+
+    return { name, namespace, kind };
+  }
+
+  // This function return the obfuscated badge url based on the namespace/kind/name triplet
+  async function getBadgeObfuscatedUrl(
+    namespace: string,
+    kind: string,
+    name: string,
+    badgeId: string,
+  ): Promise<string> {
+    const baseUrl = await options.discovery.getExternalBaseUrl('badges');
+    const salt = options.config.getString('custom.badges-backend.salt');
+    const hash = crypto
+      .createHash('sha256')
+      .update(`${kind}:${namespace}:${name}:${salt}`)
+      .digest('hex');
+    return `${baseUrl}/entity/${hash}/${badgeId}`;
+  }
 
-function getBearerToken(header?: string): string | undefined {
-  return header?.match(/Bearer\s+(\S+)/i)?.[1];
+  // This function return the badge url based on the namespace/kind/name triplet
+  async function getBadgeUrl(
+    namespace: string,
+    kind: string,
+    name: string,
+    badgeId: string,
+  ): Promise<string> {
+    const baseUrl = await options.discovery.getExternalBaseUrl('badges');
```

Comment:
We could store the baseUrl at the top and avoid the promises for this function and `getBadgeObfuscatedUrl`